### DISCUSSION TITLE
Feature: 2FA support

### DISCUSSION THREAD

**Discussion by vins345** [IceWhaleTech/CasaOS#425] [Ideas]
Hi, please add 2FA support. Login and password is not secure in 2022.

**Comment by Protektor-Desura** [comment]
You should not be exposing your home server directly to the internet. You should be using a VPN when you need to access your home server when not a home. This is the most secure way to do things and you don't have to bother with reverse proxy and expose every service/server program on your network to the internet. I recommend WireGuard or even better ZeroTier or Tailscale instead. They are both open-source and you don't need to do any port forwarding or punch a hole in  your firewall. Tailscale is built on top of Wireguard.

https://tailscale.com/
https://github.com/tailscale/tailscale

https://www.zerotier.com/
https://github.com/zerotier/ZeroTierOne

**Reply by vins345** [reply]
> You should not be exposing your home server directly to the internet. You should be using a VPN when you need to access your home server when not a home. This is the most secure way to do things and you don't have to bother with reverse proxy and expose every service/server program on your network to the internet. I recommend WireGuard or even better ZeroTier or Tailscale instead. They are both open-source and you don't need to do any port forwarding or punch a hole in your firewall. Tailscale is built on top of Wireguard.
> 
> https://tailscale.com/ https://github.com/tailscale/tailscale
> 
> https://www.zerotier.com/ https://github.com/zerotier/ZeroTierOne

Ok, but 2FA is still basic function to services like this. TrueNas have 2FA to webpanel.

**Reply by Protektor-Desura** [reply]
I'll ask again why do you need 2FA if your server and everything on it is NOT accessible from the internet except through a ZERO trust VPN? This also removes the need to set up a reverse proxy and the issue of messing with HTTP headers for some servers/apps as well. If you and those in your home are the only ones who can access your server apps why do you need 2FA? You server is shut off from the internet for all practical purposes.

**Reply by vins345** [reply]
> I'll ask again why do you need 2FA if your server and everything on it is NOT accessible from the internet except through a ZERO trust VPN? This also removes the need to set up a reverse proxy and the issue of messing with HTTP headers for some servers/apps as well. If you and those in your home are the only ones who can access your server apps why do you need 2FA? You server is shut off from the internet for all practical purposes.

So again. 2FA is basic function in truenas,synology etc. In 2022 2FA option should by exsist in all webpanels. It does not matter if expose to internet or no. Normal login and password is weak.  Why no add this optional function?

**Reply by Protektor-Desura** [reply]
Because it adds more complexity by being forced to have a domain name and set up a cert so you can do https for something that will never be accessible to the Internet.

Not everyone wants to buy a domain then go through the complexities of getting a secure cert then setting that up. It isn't even close to a couple of clicks and you're done or completely automated.

**Reply by vins345** [reply]
> Because it adds more complexity by being forced to have a domain name and set up a cert so you can do https for something that will never be accessible to the Internet.
> 
> Not everyone wants to buy a domain then go through the complexities of getting a secure cert then setting that up. It isn't even close to a couple of clicks and you're done or completely automated.

SSL can by generated localy and add for local domain configured in router ".local". No need to buy anything fo https in local network. If http enough for user then can stay with http.

**Reply by Protektor-Desura** [reply]
Self signed certs are going to throw errors telling the user not to advance because it is unsafe. So self signed certs are not an answer because browers throw errors saying it is not safe.

**Reply by vins345** [reply]
> Self signed certs are going to throw errors telling the user not to advance because it is unsafe. So self signed certs are not an answer because browers throw errors saying it is not safe.

just add them to client PC and browser will show the security lock icon.

**Reply by Protektor-Desura** [reply]
If you can not set it to click here to add your cert, then it is just going to create more support issues for having to explain things.

**Comment by DigitalRoadi3s** [comment]
if you're using file shares or Vaultwarden traffic to/from CasaOS should be encrypted.
